# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/wr/webrender/src/record.rs
# Commit: a0daec0e7301
# Full Hash: a0daec0e73011a6b972d2bc87c1cf620d569af80
# Author: Kris Taeleman <ktaeleman@mozilla.com>
# Date: 2019-09-27 04:28:18
# Regressor Bug: 1582565
# File Overlap Count: 6
# Description:
#   Bug 1582565 - Log transactions into readable format. r=jrmuizel
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D46657
# ==============================================================================

diff -r 3bd01ec4a9e9 -r a0daec0e7301 gfx/wr/webrender/src/record.rs
--- a/gfx/wr/webrender/src/record.rs	Wed Sep 25 20:46:24 2019 +0000
+++ b/gfx/wr/webrender/src/record.rs	Thu Sep 26 16:43:05 2019 +0000
@@ -61,6 +61,36 @@
     }
 }
 
+#[derive(Debug)]
+pub struct LogRecorder {
+    file: File,
+}
+
+impl LogRecorder {
+    pub fn new(dest: &PathBuf) -> LogRecorder {
+        let file = File::create(dest).unwrap();
+        LogRecorder { file }
+    }
+}
+
+impl ApiRecordingReceiver for LogRecorder {
+    fn write_msg(&mut self, _: u32, msg: &ApiMsg) {
+        let current_time = time::now_utc();
+        writeln!(self.file, "{}:{}ms - {:?}", current_time.rfc3339(), current_time.tm_nsec / 1000000, msg).unwrap();
+        match *msg {
+            ApiMsg::UpdateDocuments(_, ref msgs) => {
+                for msg in msgs {
+                    writeln!(self.file, "\tTransaction: {:?}", msg).unwrap();
+                }
+            }
+            _ => {},
+        }
+    }
+
+    fn write_payload(&mut self, _: u32, _data: &[u8]) {
+    }
+}
+
 fn should_record_transaction_msg(msgs: &TransactionMsg) -> bool {
     if msgs.generate_frame {
         return true;